File: phone-exchange-2-nodejs/REST/get_exchange_info_rest.js
import axios from 'axios';
import querystring from 'querystring';
import {PE2Response} from './pe2_response.js';

/**
 * @constant
 * @type {string}
 * @description The base URL for the live ServiceObjects Phone Exchange (PE2) API service.
 */
const LiveBaseUrl = 'https://sws.serviceobjects.com/pe2/web.svc/json/';

/**
 * @constant
 * @type {string}
 * @description The base URL for the backup ServiceObjects Phone Exchange (PE2) API service.
 */
const BackupBaseUrl = 'https://swsbackup.serviceobjects.com/pe2/web.svc/json/';

/**
 * @constant
 * @type {string}
 * @description The base URL for the trial ServiceObjects Phone Exchange (PE2) API service.
 */
const TrialBaseUrl = 'https://trial.serviceobjects.com/pe2/web.svc/json/';

/**
 * <summary>
 * Checks if a response from the API is valid by verifying that it either has no Error object
 * or the Error.TypeCode is not equal to '3'.
 * </summary>
 * <param name="response" type="Object">The API response object to validate.</param>
 * <returns type="boolean">True if the response is valid, false otherwise.</returns>
 */
const isValid = (response) => !response?.Error || response.Error.TypeCode !== '3';

/**
 * <summary>
 * Constructs a full URL for the GetExchangeInfo API endpoint by combining the base URL
 * with query parameters derived from the input parameters.
 * </summary>
 * <param name="params" type="Object">An object containing all the input parameters.</param>
 * <param name="baseUrl" type="string">The base URL for the API service (live, backup, or trial).</param>
 * <returns type="string">The constructed URL with query parameters.</returns>
 */
const buildUrl = (params, baseUrl) =>
    `${baseUrl}GetExchangeInfo?${querystring.stringify(params)}`;

/**
 * <summary>
 * Performs an HTTP GET request to the specified URL with a given timeout.
 * </summary>
 * <param name="url" type="string">The URL to send the GET request to.</param>
 * <param name="timeoutSeconds" type="number">The timeout duration in seconds for the request.</param>
 * <returns type="Promise<PE2Response>">A promise that resolves to a PE2Response object containing the API response data.</returns>
 * <exception cref="Error">Thrown if the HTTP request fails, with a message detailing the error.</exception>
 */
const httpGet = async (url, timeoutSeconds) => {
    try {
        const response = await axios.get(url, { timeout: timeoutSeconds * 1000 });
        return new PE2Response(response.data);
    } catch (error) {
        throw new Error(`HTTP request failed: ${error.message}`);
    }
};

/**
 * <summary>
 * Provides functionality to call the ServiceObjects Phone Exchange (PE2) API's GetExchangeInfo endpoint,
 * retrieving phone exchange information (e.g., carrier, line type, ported status) for a given US/Canada phone number
 * with fallback to a backup endpoint for reliability in live mode.
 * </summary>
 */
const GetExchangeInfoClient = {
    /**
     * <summary>
     * Asynchronously invokes the GetExchangeInfo API endpoint, attempting the primary endpoint
     * first and falling back to the backup if the response is invalid (Error.TypeCode == '3') in live mode.
     * </summary>
     * @param {string} PhoneNumber - The phone number to validate (e.g., "1234567890").
     * @param {string} CountryCode - 1-3 digit country calling code (e.g., "1"). Optional.
     * @param {string} Country - ISO2, ISO3, or country name (e.g., "US"). Optional.
     * @param {string} IPAddress - IPv4 address. Optional.
     * @param {string} CallerCountry - ISO2 or ISO3 code representing the caller's country. Optional.
     * @param {string} Extras - Comma-separated list of possible options. Optional.
     * @param {string} Token - Your token to use the service. Sign up for a free token at.
     * @param {string} LicenseKey - Your license key to use the service.
     * @param {boolean} isLive - Value to determine whether to use the live or trial servers.
     * @param {number} timeoutSeconds - Timeout, in seconds, for the call to the service.
     * @returns {Promise<PE2Response>} - A promise that resolves to a PE2Response object.
     */
    async invokeAsync(PhoneNumber, CountryCode, Country, IPAddress, CallerCountry, Extras,Token, LicenseKey, isLive = true, timeoutSeconds = 15) {
        const params = {
            PhoneNumber,
            CountryCode,
            Country,
            IPAddress,
            CallerCountry,
            Extras,
            Token,
            LicenseKey
        };

        const url = buildUrl(params, isLive ? LiveBaseUrl : TrialBaseUrl);
        let response = await httpGet(url, timeoutSeconds);

        if (isLive && !isValid(response)) {
            const fallbackUrl = buildUrl(params, BackupBaseUrl);
            const fallbackResponse = await httpGet(fallbackUrl, timeoutSeconds);
            return fallbackResponse;
        }
        return response;
    },

    /**
     * <summary>
     * Synchronously invokes the GetExchangeInfo API endpoint by wrapping the async call
     * and awaiting its result immediately.
     * </summary>
     * @param {string} PhoneNumber - The phone number to validate (e.g., "1234567890").
     * @param {string} CountryCode - 1-3 digit country calling code (e.g., "1"). Optional.
     * @param {string} Country - ISO2, ISO3, or country name (e.g., "US"). Optional.
     * @param {string} IPAddress - IPv4 address. Optional.
     * @param {string} CallerCountry - ISO2 or ISO3 code representing the caller's country. Optional.
     * @param {string} Extras - Comma-separated list of possible options. Optional.
     * @param {string} Token - Your token to use the service. Sign up for a free token at.
     * @param {string} LicenseKey - Your license key to use the service.
     * @param {boolean} isLive - Value to determine whether to use the live or trial servers.
     * @param {number} timeoutSeconds - Timeout, in seconds, for the call to the service.
     * @returns {PE2Response} - A PE2Response object with phone exchange details or an error.
     */
    invoke(PhoneNumber, CountryCode, Country, IPAddress, CallerCountry, Extras,Token, LicenseKey, isLive = true, timeoutSeconds = 15) {
        return (async () => await this.invokeAsync(
            PhoneNumber, CountryCode, Country, IPAddress, CallerCountry, Extras, Token, LicenseKey, isLive, timeoutSeconds
        ))();
    }
};

export { GetExchangeInfoClient, PE2Response };

File: phone-exchange-2-python/REST/get_exchange_info_rest.py
from pe2_response import PE2Response, ExchangeInfo, InternationalExchangeInfo, PortedInfo, Error
import requests

# Endpoint URLs for ServiceObjects Phone Exchange (PE2) API
primary_url = "https://sws.serviceobjects.com/pe2/web.svc/json/GetExchangeInfo?"
backup_url = "https://swsbackup.serviceobjects.com/pe2/web.svc/json/GetExchangeInfo?"
trial_url = "https://trial.serviceobjects.com/pe2/web.svc/json/GetExchangeInfo?"

def get_exchange_info(
    phone_number: str,
    license_key: str = None,
    is_live: bool = True
) -> PE2Response:
    """
    Call ServiceObjects Phone Exchange (PE2) API's GetExchangeInfo endpoint
    to retrieve phone exchange information for a given US/Canada phone number.

    Parameters:
        phone_number: The phone number to validate (e.g., "8051234567").
        license_key: Your ServiceObjects license key.
        is_live: Use live or trial servers.

    Returns:
        PE2Response: Parsed JSON response with phone exchange results or error details.

    Raises:
        RuntimeError: If the API returns an error payload.
        requests.RequestException: On network/HTTP failures (trial mode).
    """
    params = {
        "PhoneNumber": phone_number,
        "LicenseKey": license_key,
    }
    # Select the base URL: production vs trial
    url = primary_url if is_live else trial_url

    try:
        # Attempt primary (or trial) endpoint
        response = requests.get(url, params=params, timeout=10)
        response.raise_for_status()
        data = response.json()

        # If API returned an error in JSON payload, trigger fallback
        error = data.get('Error')
        if not (error is None or error.get('TypeCode') != "3"):
            if is_live:
                # Try backup URL
                response = requests.get(backup_url, params=params, timeout=10)
                response.raise_for_status()
                data = response.json()

                # If still error, propagate exception
                if 'Error' in data:
                    raise RuntimeError(f"Phone Exchange service error: {data['Error']}")
            else:
                # Trial mode error is terminal
                raise RuntimeError(f"Phone Exchange trial error: {data['Error']}")

        # Convert JSON response to PE2Response for structured access
        error = Error(**data.get("Error", {})) if data.get("Error") else None
        exchange_info_results = []
        if "ExchangeInfoResults" in data:
            for ei in data.get("ExchangeInfoResults", []):
                if isinstance(ei, dict):
                    ported_info_list = []
                    if "PortedInfo" in ei and ei.get("PortedInfo"):
                        for pi in ei.get("PortedInfo", []):
                            if isinstance(pi, dict):
                                ported_info_list.append(PortedInfo(
                                    OriginalName=pi.get("OriginalName"),
                                    OriginalLineType=pi.get("OriginalLineType"),
                                    PortedDate=pi.get("PortedDate"),
                                    LATA=pi.get("LATA")
                                ))
                    exchange_info_results.append(ExchangeInfo(
                        PhoneNumber=ei.get("PhoneNumber"),
                        Name=ei.get("Name"),
                        City=ei.get("City"),
                        State=ei.get("State"),
                        Country=ei.get("Country"),
                        LineType=ei.get("LineType"),
                        TimeZone=ei.get("TimeZone"),
                        Latitude=ei.get("Latitude"),
                        Longitude=ei.get("Longitude"),
                        SMSAddress=ei.get("SMSAddress"),
                        MMSAddress=ei.get("MMSAddress"),
                        PortedInfo=ported_info_list,
                        NoteCodes=ei.get("NoteCodes", []),
                        NoteDescriptions=ei.get("NoteDescriptions", [])
                    ))

        return PE2Response(
            ExchangeInfoResults=exchange_info_results,
            Error=error
        )

    except requests.RequestException as req_exc:
        # Network or HTTP-level error occurred
        if is_live:
            try:
                # Fallback to backup URL
                response = requests.get(backup_url, params=params, timeout=10)
                response.raise_for_status()
                data = response.json()
                if "Error" in data:
                    raise RuntimeError(f"Phone Exchange backup error: {data['Error']}") from req_exc

                error = Error(**data.get("Error", {})) if data.get("Error") else None
                exchange_info_results = []
                if "ExchangeInfoResults" in data:
                    for ei in data.get("ExchangeInfoResults", []):
                        if isinstance(ei, dict):
                            ported_info_list = []
                            if "PortedInfo" in ei and ei.get("PortedInfo"):
                                for pi in ei.get("PortedInfo", []):
                                    if isinstance(pi, dict):
                                        ported_info_list.append(PortedInfo(
                                            OriginalName=pi.get("OriginalName"),
                                            OriginalLineType=pi.get("OriginalLineType"),
                                            PortedDate=pi.get("PortedDate"),
                                            LATA=pi.get("LATA")
                                        ))
                            exchange_info_results.append(ExchangeInfo(
                                PhoneNumber=ei.get("PhoneNumber"),
                                Name=ei.get("Name"),
                                City=ei.get("City"),
                                State=ei.get("State"),
                                Country=ei.get("Country"),
                                LineType=ei.get("LineType"),
                                TimeZone=ei.get("TimeZone"),
                                Latitude=ei.get("Latitude"),
                                Longitude=ei.get("Longitude"),
                                SMSAddress=ei.get("SMSAddress"),
                                MMSAddress=ei.get("MMSAddress"),
                                PortedInfo=ported_info_list,
                                NoteCodes=ei.get("NoteCodes", []),
                                NoteDescriptions=ei.get("NoteDescriptions", [])
                            ))
                return PE2Response(
                    ExchangeInfoResults=exchange_info_results,
                    Error=error
                )
            except Exception as backup_exc:
                raise RuntimeError("Phone Exchange service unreachable on both endpoints") from backup_exc
        else:
            raise RuntimeError(f"Phone Exchange trial error: {str(req_exc)}") from req_exc